### ISSUE TITLE
Question for loss implementation

### ISSUE THREAD

**Issue by JimaoJIANG** [funkelab/cellulus#8]
Hello,
I found the loss described in the paper is distingushed from the loss implemented in the code. And in `train.py`, we calculate the loss by `embeddings_anchor` and `embeddings_reference` with out considering $d(i, j)$. In my understanding, in `dataset.zarr_dataset.py`, the `anchor_samples` and `reference_samples` are predefined. Thus, $d(i, j)$ is constant, we do not need to add it to the loss function for optimizing. I would like to ask if my understanding is correct.
Another problem that troubles me is why the mean of intra-object offset is non zero. As we have two patches $a$ and $b$, we still have the forward direction $\vec{ij}$ and backward direction $\vec{ji}$.

Thanks.

**Comment by lmanan** [comment]
Hi @JimaoJIANG . Thank you for trying out `cellulus`

1. Yes, you are right that we mention the sigmoid function in the paper 
$$\sigma(\delta) = \frac{1}{1+\exp^{-\frac{\lVert\delta\rVert^{2}_{2}}{\tau}}}, $$
but in the code [here](https://github.com/funkelab/cellulus/blob/a13ea66930ebe6ddae06d35a926decb36e139d40/cellulus/criterions/oce_loss.py#L50), we use a different function:

$$\sigma(\delta) = 1 - \exp^{-\frac{\lVert\delta\rVert ^{2}_{2}}{\tau}}. $$

The latter function has the benefit that it has a lower bound of `0` and an upper bound of `1`, whereas in the former case it would have a lower bound of `0.5`. We believe that both the functions should perform equivalently.

2. Essentially we would like that for a pair of patches $p(i)$ at location $i$, and  $p(j)$ at location $j$ with respective OCEs $r(i)$ and $r(j)$, 

$$\delta = i - j -\left( r(i) -r(j) \right), $$ becomes a zero vector by minimizing our loss function. 
In practice, we add the patch OCEs instead of subtracting them, i.e. 

$$\delta^{'} = i + r(i) -\left( j+r(j) \right).$$ 
This should have the effect that the produced OCEs are just the theoretical OCEs reflected around the y=x line.

3. Let's say we are looking at two specific kind of patches, for example, the left and right most patches illustrated in Figure 3 [here](https://openaccess.thecvf.com/content/ICCV2023/papers/Wolf_Unsupervised_Learning_of_Object-Centric_Embeddings_for_Cell_Instance_Segmentation_in_ICCV_2023_paper.pdf). 
Since for any given object, the left-most patch always appears on the left of that object and the right-most patch appears on the right of that object (and not as well to the left of the left most patch), the intra object offset should be finite and non-zero.

Hope this helps. Let us know if you have more questions. Thank you!

**Comment by JimaoJIANG** [comment]
Thank you once again for your invaluable work and insightful responses! I now have a clearer understanding of your design for the loss function, which I find quite elegant. However, I do have further questions concerning the intra-object offsets. As you have pointed out,  the left-most patches always have unique corresponding right-most patches, and the distance from the left to the right is finite and non-zero. But, when considering the direction in reverse—calculating the distance across the entire object—the right-most patch also maintains an inverse distance to the left-most patch. Consequently, the overall distance within an object have both the "left-to-right" and "right-to-left" distances.  From my understanding, given that coordinates are generated randomly, we lack precise patch correspondence within an object, leading to a non-zero intra-object distance (whereas, for inter-object offsets, the distance remains zero due to the random distribution).

 Thank you again!

**Comment by lmanan** [comment]
Just to clarify, in our formulation, we define intra and inter object offset vectors with respect to a pair of stereotypic patches present in all objects (and we do not refer to an average intra object offset vector per object, which as you alluded to, would be a zero magnitude vector). 
Essentially, we state that the expected offset vector between any two stereotypic object patches (can occur anywhere) would be equal to the intra object offset vector (as if the stereotypic patches are taken from the same object). 
Since patches are randomly extracted during training, over several iterations, the model would have learnt the relative arrangement of all pairs of stereotypic patches. So the left-most stereotypic patch of any object should map to the same location in the object centric embedding space. Similarly, the right most stereotypic patch of any object would point to another unique location, and so on. 
Does this help? Thank you!

**Comment by JimaoJIANG** [comment]
Thanks a lot! Your prompt and thoughtful response truly helped me gain a deeper understanding of this significant work. 
Once again, thank you for your invaluable contribution.
